### PR TITLE
Prefetch bugs bound extent should be attached to param or buffer

### PR DESCRIPTION
**Pull request by psuriana**
Fixed bugs when prefetching external buffers. The variable defined for bound min/extent should also attach the Buffer or Parameter.

### CODE REVIEW

**Comment by psuriana** [comment]
@dsharletg PTAL

File: src/Prefetch.cpp [hunk 1]
```diff
@@ -27,13 +27,49 @@ const Definition &get_stage_definition(const Function &f, int stage_num) {
     return f.update(stage_num - 1);
 }
 
+// Collect the bounds of all the externally referenced buffers in a stmt.
+class CollectExternalBufferBounds : public IRGraphVisitor {
```
Comment (dsharletg) [inline src/Prefetch.cpp]: Comment why this needs to be a graph visitor (or maybe it doesn't?)

File: src/Prefetch.cpp [hunk 2]
```diff
@@ -27,13 +27,49 @@ const Definition &get_stage_definition(const Function &f, int stage_num) {
...
+
+    void visit(const Call *op) {
+        IRGraphVisitor::visit(op);
+        add_buffer_bounds(op->name, op->image, op->param, (int)op->args.size());
+    }
+
+    void visit(const Variable *op) {
+        if (op->param.defined() &&
+            op->param.is_buffer() &&
+            buffers.find(op->param.name()) == buffers.end()) {
```
Comment (dsharletg) [inline src/Prefetch.cpp]: For consistency, either add this same check to the call visitor, or remove this one (don't think it's necessary). Or, comment why this one needs it and the call visitor does not.

File: src/Prefetch.cpp [hunk 3]
```diff
@@ -27,13 +27,48 @@ const Definition &get_stage_definition(const Function &f, int stage_num) {
...
+    }
+
+    void visit(const Call *op) {
+        IRVisitor::visit(op);
+        add_buffer_bounds(op->name, op->image, op->param, (int)op->args.size());
+    }
+
+    void visit(const Variable *op) {
+        if (op->param.defined() &&
+            op->param.is_buffer()) {
```
Comment (dsharletg) [inline src/Prefetch.cpp]: Minor nit: this easily fits on one line and multi-line conditions are a bit harder to read quickly.

**Comment by ronlieb** [comment]
LGTM
is this one good to go ?

**Comment by dsharletg** [comment]
Hmm seems like the buildbots didn't run. I'll test it real quick then probably merge it.

**Pull request merged**